File: server/src/main/java/de/zalando/zally/util/ast/ReverseAstBuilder.java
package de.zalando.zally.util.ast;

import com.fasterxml.jackson.annotation.JsonAnyGetter;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.core.JsonPointer;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.IdentityHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import static de.zalando.zally.util.ast.Util.PRIMITIVES;

public class ReverseAstBuilder<T> {
    private Collection<String> extensionMethodNames = new HashSet<>();

    public static class ReverseAstException extends Exception {
        ReverseAstException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    private final Deque<Node> nodes = new LinkedList<>();
    private final Map<Object, Node> objectsToNodes = new IdentityHashMap<>();
    private final Map<String, Node> pointersToNodes = new HashMap<>();

    ReverseAstBuilder(T root) {
        nodes.push(new Node(root, JsonPointers.empty(), null));
    }

    public ReverseAstBuilder<T> withExtensionMethodNames(String... names) {
        this.extensionMethodNames.addAll(Arrays.asList(names));
        return this;
    }

    /**
     * Construct a new ReverseAst instance from the root object in this builder.
     * Traverses a Swagger or OpenApi object tree and constructs a map of object nodes to meta information objects.
     *
     * @return A new ReverseAst instance.
     * @throws ReverseAstException If an error occurs during reflection.
     */
    @Nonnull
    public ReverseAst build() throws ReverseAstException {
        while (!nodes.isEmpty()) {
            Node node = nodes.pop();

            if (!PRIMITIVES.contains(node.object.getClass())) {
                Collection<Node> children;
                if (node.object instanceof Map) {
                    children = handleMap((Map<?, ?>) node.object, node.pointer, node.marker);
                } else if (node.object instanceof List) {
                    children = handleList((List<?>) node.object, node.pointer, node.marker);
                } else if (node.object instanceof Set) {
                    children = handleSet((Set<?>) node.object, node.pointer, node.marker);
                } else if (node.object instanceof Object[]) {
                    children = handleArray((Object[]) node.object, node.pointer, node.marker);
                } else {
                    children = handleObject(node.object, node.pointer, node.marker);
                }
                for (Node child : children) {
                    nodes.push(child);
                }
                node.setChildren(children);
            }
            if (!node.skip) {
                objectsToNodes.put(node.object, node);
                pointersToNodes.put(node.pointer.toString(), node);
            }
        }
        return new ReverseAst(objectsToNodes, pointersToNodes);
    }

    private Deque<Node> handleMap(Map<?, ?> map, JsonPointer pointer, Marker defaultMarker) {
        final Deque<Node> nodes = new LinkedList<>();
        final Marker marker = getMarker(map).orElse(defaultMarker);

        for (Map.Entry<?, ?> entry : map.entrySet()) {
            Object key = entry.getKey();
            Object value = entry.getValue();
            if (key instanceof String && value != null) {
                JsonPointer newPointer = pointer.append(JsonPointers.escape((String) key));
                nodes.push(new Node(value, newPointer, marker));
            }
        }
        return nodes;
    }

    private Deque<Node> handleList(List<?> list, JsonPointer pointer, Marker marker) {
        return handleArray(list.toArray(), pointer, marker);
    }

    private Deque<Node> handleSet(Set<?> set, JsonPointer pointer, Marker marker) {
        return handleArray(set.toArray(), pointer, marker);
    }

    private Deque<Node> handleArray(Object[] objects, JsonPointer pointer, Marker marker) {
        final Deque<Node> nodes = new LinkedList<>();

        for (int i = 0; i < objects.length; i++) {
            Object value = objects[i];
            if (value != null) {
                JsonPointer newPointer = pointer.append(JsonPointers.escape(String.valueOf(i)));
                nodes.push(new Node(value, newPointer, marker));
            }
        }
        return nodes;
    }

    private Deque<Node> handleObject(Object object, JsonPointer pointer, Marker defaultMarker) throws ReverseAstException {
        final Deque<Node> nodes = new LinkedList<>();
        final Marker marker = getMarker(object).orElse(defaultMarker);

        for (Method m : object.getClass().getDeclaredMethods()) {
            String name = m.getName();
            // Find all public getter methods.
            if (isPublicGetterMethod(m)) {
                try {
                    Object value = m.invoke(object);
                    if (value != null) {
                        if (m.isAnnotationPresent(JsonAnyGetter.class)) {
                            // A `JsonAnyGetter` method is simply a wrapper for nested properties.
                            // We must not use the method name but re-use the current pointer.
                            nodes.push(new Node(value, pointer, marker, /* skip */true));
                        } else {
                            JsonPointer newPointer = pointer.append(JsonPointers.escape(m));
                            nodes.push(new Node(value, newPointer, marker));
                        }
                    }
                } catch (ReflectiveOperationException e) {
                    String message = String.format("Error invoking %s on %s at path %s", name, object.getClass(), pointer);
                    throw new ReverseAstException(message, e);
                }
            }
        }
        return nodes;
    }

    private boolean isPublicGetterMethod(Method m) {
        return m.getName().startsWith("get")
            && m.getParameterCount() == 0
            && Modifier.isPublic(m.getModifiers())
            && !m.isAnnotationPresent(JsonIgnore.class);
    }

    private Optional<Marker> getMarker(Map<?, ?> map) {
        return Optional
            .ofNullable(getVendorExtensions(map, Marker.TYPE_X_ZALLY_IGNORE))
            .map(values -> new Marker(Marker.TYPE_X_ZALLY_IGNORE, values));
    }

    private Optional<Marker> getMarker(Object object) throws ReverseAstException {
        return Optional
            .ofNullable(getVendorExtensions(object, Marker.TYPE_X_ZALLY_IGNORE))
            .map(values -> new Marker(Marker.TYPE_X_ZALLY_IGNORE, values));
    }

    @Nullable
    private Collection<String> getVendorExtensions(Object object, String extensionName) throws ReverseAstException {
        if (object instanceof Map) {
            return getVendorExtensions((Map) object, extensionName);
        }
        for (Method m : object.getClass().getDeclaredMethods()) {
            if (extensionMethodNames.contains(m.getName())) {
                try {
                    Object extensions = m.invoke(object);
                    if (extensions instanceof Map) {
                        return getVendorExtensions((Map) extensions, extensionName);
                    }
                } catch (ReflectiveOperationException e) {
                    throw new ReverseAstException("Error getting extensions.", e);
                }
            }
        }
        return null;
    }

    @Nullable
    private Collection<String> getVendorExtensions(Map<?, ?> map, String extensionName) {
        if (map.containsKey(extensionName)) {
            Object value = ((Map) map).get(extensionName);
            if (value instanceof String) {
                return Collections.singleton((String) value);
            }
            if (value instanceof Collection) {
                return ((Collection<?>) value).stream().map(Object::toString).collect(Collectors.toSet());
            }
        }
        return null;
    }
}

File: server/src/main/java/de/zalando/zally/util/ast/ReverseAstBuilder.kt
package de.zalando.zally.util.ast

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.core.JsonPointer
import java.lang.reflect.Method
import java.lang.reflect.Modifier
import java.util.Deque
import java.util.IdentityHashMap
import java.util.LinkedList
import kotlin.Comparator
import kotlin.collections.HashMap
import kotlin.collections.HashSet
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.set

class ReverseAstBuilder<T> internal constructor(root: T) {
    private val extensionMethodNames = HashSet<String>()

    private val nodes = LinkedList<Node>()
    private val objectsToNodes = IdentityHashMap<Any, Node>()
    private val pointersToNodes = HashMap<String, Node>()

    class ReverseAstException internal constructor(message: String, cause: Throwable) : Exception(message, cause)

    init {
        nodes.push(Node(root as Any, JsonPointers.EMPTY, null))
    }

    fun withExtensionMethodNames(vararg names: String): ReverseAstBuilder<T> {
        this.extensionMethodNames.addAll(listOf(*names))
        return this
    }

    /**
     * Construct a new ReverseAst instance from the root object in this builder.
     * Traverses a Swagger or OpenApi object tree and constructs a map of object nodes to meta information objects.
     *
     * @return A new ReverseAst instance.
     * @throws ReverseAstException If an error occurs during reflection.
     */
    @Throws(ReverseAstException::class)
    fun build(): ReverseAst {
        while (!nodes.isEmpty()) {
            val node = nodes.pop()
            if (objectsToNodes.containsKey(node.`object`)) {
                continue
            }
            if (!Util.PRIMITIVES.contains(node.`object`.javaClass)) {
                val children = when {
                    node.`object` is Map<*, *> -> handleMap(node.`object`, node.pointer, node.marker)
                    node.`object` is List<*> -> handleList(node.`object`, node.pointer, node.marker)
                    node.`object` is Set<*> -> handleSet(node.`object`, node.pointer, node.marker)
                    node.`object` is Array<*> -> handleArray(node.`object`, node.pointer, node.marker)
                    else -> handleObject(node.`object`, node.pointer, node.marker)
                }
                for (child in children) {
                    nodes.push(child)
                }
                node.children.addAll(children)
            }
            if (!node.skip) {
                objectsToNodes[node.`object`] = node
                pointersToNodes[node.pointer.toString()] = node
            }
        }
        return ReverseAst(objectsToNodes, pointersToNodes)
    }

    private fun handleMap(map: Map<*, *>, pointer: JsonPointer, defaultMarker: Marker?): Deque<Node> {
        val nodes = LinkedList<Node>()
        val marker = getMarker(map) ?: defaultMarker

        for ((key, value) in map) {
            if (key is String && value != null) {
                val newPointer = pointer.append(JsonPointers.escape(key))
                nodes.push(Node(value, newPointer, marker))
            }
        }
        return nodes
    }

    private fun handleList(list: List<*>, pointer: JsonPointer, marker: Marker?): Deque<Node> {
        return handleArray(list.toTypedArray(), pointer, marker)
    }

    private fun handleSet(set: Set<*>, pointer: JsonPointer, marker: Marker?): Deque<Node> {
        return handleArray(set.toTypedArray(), pointer, marker)
    }

    private fun handleArray(objects: Array<*>, pointer: JsonPointer, marker: Marker?): Deque<Node> {
        val nodes = LinkedList<Node>()

        for (i in objects.indices) {
            val value = objects[i]
            if (value != null) {
                val newPointer = pointer.append(JsonPointers.escape(i.toString()))
                nodes.push(Node(value, newPointer, marker))
            }
        }
        return nodes
    }

    @Throws(ReverseAstException::class)
    private fun handleObject(`object`: Any, pointer: JsonPointer, defaultMarker: Marker?): Deque<Node> {
        val nodes = LinkedList<Node>()
        val marker = getMarker(`object`) ?: defaultMarker

        for (m in traversalMethods(`object`.javaClass)) {
            val name = m.name
            try {
                val value = m.invoke(`object`)
                if (value != null) {
                    if (m.isAnnotationPresent(JsonAnyGetter::class.java)) {
                        // A `JsonAnyGetter` method is simply a wrapper for nested properties.
                        // We must not use the method name but re-use the current pointer.
                        nodes.push(Node(value, pointer, marker, /* skip */true))
                    } else {
                        val newPointer = pointer.append(JsonPointers.escape(m))
                        nodes.push(Node(value, newPointer, marker))
                    }
                }
            } catch (e: ReflectiveOperationException) {
                val message = String.format("Error invoking %s on %s at path %s", name, `object`.javaClass, pointer)
                throw ReverseAstException(message, e)
            }
        }
        return nodes
    }

    private fun getMarker(map: Map<*, *>): Marker? =
        getVendorExtensions(map, Marker.TYPE_X_ZALLY_IGNORE)
            ?.let { Marker(Marker.TYPE_X_ZALLY_IGNORE, it) }

    @Throws(ReverseAstException::class)
    private fun getMarker(`object`: Any): Marker? =
        getVendorExtensions(`object`, Marker.TYPE_X_ZALLY_IGNORE)
            ?.let { Marker(Marker.TYPE_X_ZALLY_IGNORE, it) }

    @Throws(ReverseAstBuilder.ReverseAstException::class)
    private fun getVendorExtensions(`object`: Any, extensionName: String): Collection<String>? {
        if (`object` is Map<*, *>) {
            return getVendorExtensions(`object`, extensionName)
        }
        for (m in traversalMethods(`object`.javaClass)) {
            if (extensionMethodNames.contains(m.name)) {
                try {
                    val extensions = m.invoke(`object`)
                    if (extensions is Map<*, *>) {
                        return getVendorExtensions(extensions, extensionName)
                    }
                } catch (e: ReflectiveOperationException) {
                    throw ReverseAstException("Error getting extensions.", e)
                }
            }
        }
        return null
    }

    private fun getVendorExtensions(map: Map<*, *>, extensionName: String): Collection<String>? {
        if (map.containsKey(extensionName)) {
            val value = map[extensionName]
            if (value is String) {
                return setOf(value)
            }
            if (value is Collection<*>) {
                return value.map { it.toString() }.toSet()
            }
        }
        return null
    }

    companion object {
        fun traversalMethods(clazz: Class<*>) = clazz
            .methods
            .filter {
                it.name.startsWith("get") &&
                    it.name != "getClass" &&
                    it.name != "getDeclaringClass" &&
                    it.parameterCount == 0 &&
                    Modifier.isPublic(it.modifiers) &&
                    !it.isAnnotationPresent(JsonIgnore::class.java)
            }
            .sortedWith(Comparator
                .comparing { method: Method -> method.name == "getPaths" }
                .thenComparing { method: Method -> method.name })
    }
}
